pyteal/ast/subroutine.py

```diff
@@ -374,15 +425,138 @@ def name(self) -> str:
         return self.subroutine.name()
 
     def type_of(self):
-        return self.subroutine.getDeclaration().type_of()
+        return self.subroutine.get_declaration().type_of()
 
     def has_return(self):
-        return self.subroutine.getDeclaration().has_return()
+        return self.subroutine.get_declaration().has_return()
 
 
 SubroutineFnWrapper.__module__ = "pyteal"
 
 
+class ABIReturnSubroutine:
+    """Used to create a PyTeal Subroutine (returning an ABI value) from a python function.
+
+    This class is meant to be used as a function decorator. For example:
+
+        .. code-block:: python
+
+            @ABIReturnSubroutine
+            def abi_sum(toSum: abi.DynamicArray[abi.Uint64], *, output: abi.Uint64) -> Expr:
+                i = ScratchVar(TealType.uint64)
+                valueAtIndex = abi.Uint64()
+                return Seq(
+                    output.set(0),
+                    For(i.store(Int(0)), i.load() < toSum.length(), i.store(i.load() + Int(1))).Do(
+                        Seq(
+                            toSum[i.load()].store_into(valueAtIndex),
+                            output.set(output.get() + valueAtIndex.get()),
+                        )
+                    ),
+                )
+
+            program = Seq(
+                (to_sum_arr := abi.DynamicArray(abi.Uint64TypeSpec())).decode(
+                    Txn.application_args[1]
+                ),
+                (res := abi.Uint64()).set(abi_sum(to_sum_arr)),
+                abi.MethodReturn(res),
+                Int(1),
+            )
+    """
+
+    def __init__(
+        self,
+        fn_implementation: Callable[..., Expr],
+    ) -> None:
+        self.output_kwarg_info: Optional[
+            _OutputKwArgInfo
+        ] = self._output_name_type_from_fn(fn_implementation)
+
+        internal_subroutine_ret_type = (
+            TealType.none
+            if self.output_kwarg_info is None
+            else self.output_kwarg_info.abi_type.storage_type()
+        )
```

Comment:
```suggestion
        internal_subroutine_ret_type = None
        if self.output_kwarg_info:
            internal_subroutine_ret_type = self.output_kwarg_info.abi_type.storage_type()
```
Totally optional, but you may find a bit more readable. Given that `_output_name_type_from_fn()` is guaranteed to return `None` or something `truthy` this is equivalent.

Reply:
(ok, it's ever so slightly slower than the original)

Reply:
I just remembered, `mypy` may not like that approach and you might need to specify the type. IE:
```python
user_defined_annotations: type = None # or something like that
if ...
```

Reply:
```python
        internal_subroutine_ret_type = TealType.none
        if self.output_kwarg_info:
            internal_subroutine_ret_type = (
                self.output_kwarg_info.abi_type.storage_type()
            )
```
current implementation resolves it